### PR TITLE
Add support for azure stack client authentication

### PR DESCRIPTION
**Pull request by PrasadG193**
## Change Overview

Add support for azure stack client authentication

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

Manually tested on public azure as well as azure stack

### CODE REVIEW

File: pkg/blockstorage/azure/azuredisk.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 	"strings"
 	"time"
 
-	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
```
Comment (PrasadG193) [inline pkg/blockstorage/azure/azuredisk.go:10]: As per the docs - https://docs.microsoft.com/en-us/azure-stack/user/azure-stack-version-profiles-go?view=azs-2008
`2019-03-01` api version is compatible with the Azure stack as well as public Azure instances. I have also validated that with manual testing

File: pkg/blockstorage/helpers.go [hunk 2]
```diff
@@ -33,8 +33,11 @@ const (
 	AzureCientID               = "AZURE_CLIENT_ID"
 	AzureClentSecret           = "AZURE_CLIENT_SECRET"
 	AzureResurceGroup          = "AZURE_RESOURCE_GROUP"
+	AzureResurceMgrEndpoint    = "AZURE_RESOURCE_MANAGER_ENDPOINT"
```
Comment (onkarbhat) [inline pkg/blockstorage/helpers.go:36]: Should we add the new configs in this PR to https://github.com/kanisterio/kanister/blob/ab6010860929c1702fb94c238607a6115bcc621a/pkg/blockstorage/azure/client_test.go#L39 ? The test is currently being skipped . File a ticket to track  it ?

File: pkg/blockstorage/helpers.go [hunk 3]
```diff
@@ -33,8 +33,11 @@ const (
...
 	AzureMigrateStorageAccount = "AZURE_MIGRATE_STORAGE_ACCOUNT_NAME"
 	AzureMigrateStorageKey     = "AZURE_MIGRATE_STORAGE_ACCOUNT_KEY"
+	AzureActiveDirEndpoint     = "AZURE_AD_ENDPOINT"
```
Comment (onkarbhat) [inline pkg/blockstorage/helpers.go:39]: Update this test with the new configs ? 
https://github.com/kanisterio/kanister/blob/ab6010860929c1702fb94c238607a6115bcc621a/pkg/blockstorage/blockstorage_test.go#L301

File: pkg/blockstorage/azure/client.go [hunk 4]
```diff
@@ -98,10 +108,19 @@ func getAuthorizer(env azure.Environment, config map[string]string) (*autorest.B
...
 
 	credConfig := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+	if aDDEndpoint, ok := config[blockstorage.AzureActiveDirEndpoint]; ok {
+		credConfig.AADEndpoint = aDDEndpoint
+	}
+
+	if aDDResourceID, ok := config[blockstorage.AzureActiveDirResourceID]; ok {
+		credConfig.Resource = aDDResourceID
+	}
+
```
Comment (onkarbhat) [inline pkg/blockstorage/azure/client.go:118]: Is there a PR that will setup this new config ?

**Review by onkarbhat — COMMENTED** [review]
@PrasadG193 -  Are there existing tests that will test this change ?   I've posted a few more questions.

File: pkg/blockstorage/azure/client.go [hunk 5]
```diff
@@ -98,10 +108,19 @@ func getAuthorizer(env azure.Environment, config map[string]string) (*autorest.B
 	}
 
 	credConfig := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+	if aDDEndpoint, ok := config[blockstorage.AzureActiveDirEndpoint]; ok {
```
[the remark this answers is not in the document]
Comment (michaelcourcy) [reply pkg/blockstorage/azure/client.go:111]: @onkarbhat it's applicable 
```
az cloud show --name AzureCloud  | jq ".endpoints.activeDirectoryResourceId" 
"https://management.core.windows.net/"
az cloud show --name stackpoc  | jq ".endpoints.activeDirectoryResourceId" 
"https://management.stackpoc.com/71fb132f-bfbf-4e60-8cb6-923303347e19"
```

But you're right we should default to the public azure cloud if not specified. @PrasadG193 is it default that you specified in helm option or by other way ?

[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/blockstorage/azure/client.go:111]: > Is the AzureActiveDirEndpoint applicable to public azure too 

No. We don't need to set AzureActiveDirEndpoint if you are using public cloud. Go sdks use default global endpoint if not specified

> Wondering if this should be added all the time like the way it is setup currently in the PR, or it should be added conditionally for azure stack only ?

This is applicable for only azure stack

File: pkg/blockstorage/helpers.go (hunk 2 above)
Comment (PrasadG193) [reply pkg/blockstorage/helpers.go:36]: Yes, all azure tests are skipped as of now. Will search if a ticket exists or will file one

File: pkg/blockstorage/azure/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (onkarbhat) [reply pkg/blockstorage/azure/client.go:111]: Thanks for clarifying Prasad.

File: pkg/blockstorage/helpers.go (hunk 2 above)
Comment (onkarbhat) [reply pkg/blockstorage/helpers.go:36]: 👍🏼

File: pkg/blockstorage/helpers.go (hunk 3 above)
Comment (onkarbhat) [reply pkg/blockstorage/helpers.go:39]: Discussed this with @PrasadG193  -  He will address this in a follow up PR.

**Review by onkarbhat — APPROVED** [review]
LGTM

**Pull request merged**